// nitropack/dist/index.mjs
export { G as GLOB_SCAN_PATTERN, b as build, c as copyPublicAssets, f as createDevServer, a as createNitro, g as defineNitroConfig, l as loadOptions, p as prepare, h as prerender, e as scanAPI, s as scanHandlers, d as scanMiddleware, w as writeTypes } from './chunks/prerender.mjs';
import 'worker_threads';
import 'fs';
import 'chokidar';
import 'perfect-debounce';
import 'h3';
import 'http-proxy';
import 'listhen';
import 'serve-static';
import 'pathe';
import 'connect';
import 'ufo';
import 'hookable';
import 'unimport';
import 'consola';
import 'c12';
import 'klona/full';
import 'defu';
import 'module';
import 'fs-extra';
import 'jiti';
import 'chalk';
import 'dot-prop';
import 'url';
import 'archiver';
import 'globby';
import 'pkg-types';
import 'rollup';
import 'pretty-bytes';
import 'gzip-size';
import 'std-env';
import '@nuxt/devalue';
import 'rollup-plugin-terser';
import '@rollup/plugin-commonjs';
import '@rollup/plugin-node-resolve';
import '@rollup/plugin-alias';
import '@rollup/plugin-json';
import '@rollup/plugin-replace';
import '@rollup/plugin-virtual';
import '@rollup/plugin-wasm';
import '@rollup/plugin-inject';
import 'rollup-plugin-visualizer';
import 'unenv';
import 'mlly';
import 'unimport/unplugin';
import '@vercel/nft';
import 'semver';
import 'etag';
import 'mime';
import 'hasha';
import 'table';
import 'is-primitive';
import 'esbuild';
import '@rollup/pluginutils';
