// @nuxt/nitro/dist/runtime/entries/lambda.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = void 0;

require("#polyfill");

var _ufo = require("ufo");

var _server = require("../server");

const handler = async function handler2(event, context) {
  const url = (0, _ufo.withQuery)(event.path || event.rawPath, event.queryStringParameters);
  const method = event.httpMethod || event.requestContext?.http?.method || "get";

  if ("cookies" in event) {
    event.headers.cookie = event.cookies.join(",");
  }

  const r = await (0, _server.localCall)({
    event,
    url,
    context,
    headers: normalizeIncomingHeaders(event.headers),
    method,
    query: event.queryStringParameters,
    body: event.body
  });
  return {
    statusCode: r.status,
    headers: normalizeOutgoingHeaders(r.headers),
    body: r.body.toString()
  };
};

exports.handler = handler;

function normalizeIncomingHeaders(headers) {
  return Object.fromEntries(Object.entries(headers).map(([key, value]) => [key.toLowerCase(), value]));
}

function normalizeOutgoingHeaders(headers) {
  return Object.fromEntries(Object.entries(headers).map(([k, v]) => [k, Array.isArray(v) ? v.join(",") : v]));
}
